src/buildstream/storage/directory.py

```diff
@@ -16,328 +17,499 @@
 #
 #  Authors:
 #        Jim MacArthur <jim.macarthur@codethink.co.uk>
+#        Tristan van Berkom <tristan.vanberkom@codethink.co.uk>
 
 """
-Directory
-=========
-
-This is a virtual Directory class to isolate the rest of BuildStream
-from the backing store implementation.  Sandboxes are allowed to read
-from and write to the underlying storage, but all others must use this
-Directory class to access files and directories in the sandbox.
-
-See also: :ref:`sandboxing`.
-
+Directory - Interfacing with files
+==================================
+The Directory class is given to plugins by way of the :class:`.Sandbox`
+or in some other instances, and allows plugins to interface with files
+and directory hierarchies owned by BuildStream.
+
+
+.. _directory_path:
+
+Paths
+-----
+The path argument to directory functions depict a relative path. Path elements are
+separated with the ``/`` character regardless of the platform. Both ``.`` and ``..`` entries
+are permitted. Absolute paths are not permitted, as such it is illegal to specify a path
+with a leading ``/`` character.
+
+Directory objects represent a directory entry within the context of a *directory tree*,
+and the directory returned by
+:func:`Sandbox.get_virtual_directory() <buildstream.sandbox.Sandbox.get_virtual_directory>`
+is the root of the sandbox's *directory tree*. Attempts to escape the root of a *directory tree*
+using ``..`` entries will not result in an error, instead ``..`` entries which cross the
+root boundary will be evaluated as the root directory.
```

Comment:
Maybe mention that this matches POSIX behavior of filesystem roots?

Reply:
Appended to this paragraph:

> This behavior matches POSIX behavior of filesystem root directories.